README.md
## RcppUTS [![Build Status](https://travis-ci.org/eddelbuettel/rcpputs.svg)](https://travis-ci.org/eddelbuettel/rcpputs) [![License](https://eddelbuettel.github.io/badges/GPL2+.svg)](http://www.gnu.org/licenses/gpl-2.0.html) 

Rcpp bindings for UTS (Unevenly Spaced Time Series)

### What is UTS?

UTS is a library by [Andreas Eckner](http://eckner.com/index.html) described and made
available via his [research page](http://eckner.com/research.html) which also contains
a [zip archive](http://eckner.com/papers/uts_algorithms.zip). 

### What is RcppUTS?

This package wraps (the older library version of) UTS for use by R via
[Rcpp](http://dirk.eddelbuettel.com/code/rcpp.html).

Some examples are included, based on the example in the UTS library. Below we show the
exponential-moving average operator, available condition on the `last`, `next` (ie 
current) observation, or a linear interpolation.

![EMA](https://eddelbuettel.github.io/rcpputs/figures/emaPlot.png)\

Also available is a standard moving average with a fixed 'before' and 'after' window length.

![SMA](https://eddelbuettel.github.io/rcpputs/figures/smaPlot.png)\

Both figures are also produced by the respective `example()` commands associate with
the corresponding help files.

### What else?

Well once I told Andreas I was putting together a quick package, he got motivated too and
released [an updated GitHUb repo for uts](https://github.com/andreas50/uts) as well
[one for utsOperators](https://github.com/andreas50/utsOperators). You probably want to
watch those spaces.  His [utsOperators](https://github.com/andreas50/utsOperators) is
closer to what we do here with a focus on SMA, EMA and rolling operators.

### Status

The package builds and checks cleanly.  

More functionality could be added, of course. Contributions are welcome.

### Author

Dirk Eddelbuettel for the package and interface, Andreas Eckner for UTS.

### License

GPL (>= 2)




ChangeLog
2018-06-13  Dirk Eddelbuettel  <edd@debian.org>

	* src/emaWrapper.cpp: Update plot.txt in example
	* man/EMAnext.Rd: Idem

	* src/smaWrapper.cpp: Update example
	* man/SMAnext.Rd: Idem

2018-06-10  Dirk Eddelbuettel  <edd@debian.org>

	* src/emaWrapper.cpp: Update example
	* man/EMAnext.Rd: Idem

	* src/smaWrapper.cpp: Update example
	* man/SMAnext.Rd: Idem

	* .travis.yml (install): Install suggested package xts as needed
	for examples

	* README.md: Added examples

2018-06-06  Dirk Eddelbuettel  <edd@debian.org>

	* src/rollingWrapper.cpp: Rolling functions
	* man/rollingCentralMoment.Rd: Documentation

	* README.md: Added
	* rcpputs.Rproj: Added
	* .travis.yml: Added

2018-06-05  Dirk Eddelbuettel  <edd@debian.org>

	* src/emaWrapper.cpp: EMA functions
	* man/EMAnext.Rd: Documentation

	* src/smaWrapper.cpp: SMA functions
	* man/SMAnext.Rd: Documentation

	* src/test.cpp (utsExample): Added docs and example
	* man/utsExample.Rd: Idem

2018-06-04  Dirk Eddelbuettel  <edd@debian.org>

	* src/test.cpp: Initial commit with original example


src/test.cpp
// Copyright 2012-2017 by Andreas Eckner
// License GPL-2 | GPL-3

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

extern "C" {
#include "ema.h"
#include "sma.h"
#include "rolling.h"
}

#include <Rcpp.h>


// Print nicely formatted observation times and values for an unevenly spaced time series
void print_uts(double values[], double times[], int n)
{
  // values     ... array of observation values
  // times      ... array of observation times
  // n          ... number of observations, i.e. length of 'values' and 'times'  
  
  // Print header
  Rprintf("-------------\n");
  Rprintf("Time    Value\n");
  Rprintf("-------------\n");

  for (int i=0; i < n; i++)
    Rprintf("%.1f %9.2f\n", times[i], values[i]); 
}


//' The UTS library by Andreas Eckner provides algorithms for unevenly
//' spaced time-series data.  This package brings a few of them to R.
//' This function shows the original example.
//' @title Irregularly spaced time series example
//' @return Nothing
//' @author Dirk Eddelbuettel for the package, Andreas Eckner for the
//' underlying code.
//' @examples
//' utsExample()
// [[Rcpp::export]]
void utsExample() {
  // Define sample time series
  double values[] = {0, 2, 4, 6, 8, 10};
  double times[] = {0, 1, 1.2, 2.3, 2.9, 5};
  int n = sizeof(values) / sizeof(double);
  double out[n];
  Rprintf("Input time series X\n");
  print_uts(values, times, n);
  
  // Parameters for the demo
  double width_before = 2.5, width_after = 1, width_before_long = 1000;
  double tau = 1.5, tau_long = 1000;

  /*
    Basic Rolling Time Series Operators
  */
  Rprintf("\n\n##### Basic Rolling Time Series Operators #####\n");

  // rolling numer of observations
  rolling_num_obs(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_num_obs(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);

  // rolling sum
  rolling_sum(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_sum(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);

  // rolling product
  rolling_product(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_product(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);

  // rolling average
  rolling_mean(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_mean(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);

  // rolling median
  rolling_median(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_median(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);
  
  // rolling maximum
  rolling_max(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_max(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);
  
  // rolling minimum
  rolling_min(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_min(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);
  
  // rolling standard deviation
  rolling_sd(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_sd(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);
  
  // rolling variance
  rolling_var(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nrolling_var(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);


  /*
    Simple Moving Averages (SMAs)
  */
  Rprintf("\n\n##### Simple Moving Averages (SMAs) #####\n");

  // SMA with last-point interpolation
  sma_last(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nSMA_last(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);
  
  // SMA with next-point interpolation
  sma_next(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nSMA_next(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);

  // SMA with linear interpolation
  sma_linear(values, times, &n, out, &width_before, &width_after);
  Rprintf("\nSMA_linear(X, %.1f, %.1f)\n", width_before, width_after);
  print_uts(out, times, n);
  
  // SMA with slow time decay
  sma_linear(values, times, &n, out, &width_before_long, &width_after);
  Rprintf("\nSMA_linear(X, %.1f, %.1f) ... a SMA with a long rolling time window produces nearly constant output\n", width_before_long, width_after);
  print_uts(out, times, n);
  

  /*
    Explonential Moving Averages (EMAs)
  */
  Rprintf("\n\n##### Exponential Moving Averages (EMAs) #####\n");

  // EMA with last-point interpolation
  ema_last(values, times, &n, out, &tau);
  Rprintf("\nEMA_last(X, %.1f)\n", tau);
  print_uts(out, times, n);

  // EMA with next-point interpolation
  ema_next(values, times, &n, out, &tau);
  Rprintf("\nEMA_next(X, %.1f)\n", tau);
  print_uts(out, times, n);

  // EMA with linear interpolation
  ema_linear(values, times, &n, out, &tau);
  Rprintf("\nEMA_linear(X, %.1f)\n", tau);
  print_uts(out, times, n);
  
  // EMA with slow time decay
  ema_linear(values, times, &n, out, &tau_long);
  Rprintf("\nEMA_linear(X, %.1f) ... an EMA with a slow time decay produces nearly constant output\n", tau_long);
  print_uts(out, times, n);

  // Wait for key pressed before exiting
  //printf("\nPress <ENTER> to exit the program.\n");
  //getchar();
  //return 0;
}


src/emaWrapper.cpp
#include <Rcpp.h>

extern "C" {
#include "ema.h"
}

//' The UTS library by Andreas Eckner provides algorithms for unevenly
//' spaced time-series data.  This package brings a few of them to R.
//' The functions describe here offer exponentially-decaying weighted moving
//' average, or EMA, for short. Three variants are provides considering
//' the last or next observation relative to time \sQuote{t}, as well as
//' linear interpolation between them.
//' @title EMA functions for unevenly spaced time series
//' @param times A Datetime vector
//' @param values A numeric vector
//' @param tau A double with the decay factor
//' @return A numeric vector with EMA-weighted values.
//' package at the given position is available.
//' @author Dirk Eddelbuettel for the package, Andreas Eckner for the
//' underlying code.
//' @examples
//' if (requireNamespace("xts", quietly=TRUE)) {
//'     suppressMessages(library(xts))
//'     times <- ISOdatetime(2010, 1, 2, 8, 30, 0) + c(0, 1.0, 1.2, 2.3, 2.9, 5.0)
//'     values <- seq(0, 10, by=2)
//'     plot(xts(values, order.by=times), type="b",
//'          main="Series and last/next/linear EMAs",
//'          major.ticks="auto", grid.ticks.on="auto")
//'     lines(xts(EMAlast(times,values, 1),  values, order.by=times),
//'           type="b", col="lightblue")
//'     lines(xts(EMAnext(times,values, 1),  values, order.by=times),
//'           type="b", col="darkblue")
//'     lines(xts(EMAlinear(times,values, 1),  values, order.by=times),
//'           type="b", col="mediumblue")
//'     addLegend("topleft", legend.names=c("series", "EMAlast", "EMAnext", "EMAlinear"),
//'               lty=rep(1,4), lwd=rep(1,4),
//'               col=c("black", "lightblue", "darkblue", "mediumblue"))
//' }
// [[Rcpp::export]]
Rcpp::NumericVector EMAnext(Rcpp::DatetimeVector times,
                            Rcpp::NumericVector values,
                            const double tau) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  ema_next(values.begin(), times.begin(), &n, res.begin(), const_cast<double*>(&tau));
  return res;
}

//' @rdname EMAnext
// [[Rcpp::export]]
Rcpp::NumericVector EMAlast(Rcpp::DatetimeVector times,
                            Rcpp::NumericVector values,
                            const double tau) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  ema_last(values.begin(), times.begin(), &n, res.begin(), const_cast<double*>(&tau));
  return res;
}

//' @rdname EMAnext
// [[Rcpp::export]]
Rcpp::NumericVector EMAlinear(Rcpp::DatetimeVector times,
                              Rcpp::NumericVector values,
                              const double tau) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  ema_linear(values.begin(), times.begin(), &n, res.begin(), const_cast<double*>(&tau));
  return res;
}


src/smaWrapper.cpp
#include <Rcpp.h>

extern "C" {
#include "sma.h"
}

//' The UTS library by Andreas Eckner provides algorithms for unevenly
//' spaced time-series data.  This package brings a few of them to R.
//' The functions describe here offer simple moving
//' average, or SMA, for short. Three variants are provides considering
//' the last or next observation relative to time \sQuote{t}, as well as
//' linear interpolation between them.
//' @title SMA functions for unevenly spaced time series
//' @param times A Datetime vector
//' @param values A numeric vector
//' @param widthbefore A double with the preceding observation width
//' @param widthafter gvA double with the subsequent observation width
//' @return A numeric vector with SMA-weighted values.
//' package at the given position is available.
//' @author Dirk Eddelbuettel for the package, Andreas Eckner for the
//' underlying code.
//' @examples
//' if (requireNamespace("xts", quietly=TRUE)) {
//'     suppressMessages(library(xts))
//'     times <- ISOdatetime(2018, 6, 7, 8, 30, 0) + c(0, 1.0, 1.2, 2.3, 2.9, 5.0)
//'     values <- seq(0, 10, by=2)
//'     plot(xts(values, order.by=times), type="b",
//'          main="Series and last/next/linear SMAs",
//'          major.ticks="auto", grid.ticks.on="auto")
//'     lines(xts(SMAlast(times,values, 2.5, 1), values, order.by=times), 
//'          type="b", col="lightblue")
//'     lines(xts(SMAnext(times,values, 2.5, 1),  values, order.by=times),
//'          type="b", col="darkblue")
//'     lines(xts(SMAlinear(times,values, 2.5, 1),  values, order.by=times),
//'          type="b", col="mediumblue")
//'     addLegend("topleft", legend.names=c("series", "SMAlast", "SMAnext", "SMAlinear"),
//'               lty=rep(1,4), lwd=rep(1,4),
//'               col=c("black", "lightblue", "darkblue", "mediumblue"))
//' }
// [[Rcpp::export]]
Rcpp::NumericVector SMAnext(Rcpp::DatetimeVector times,
                            Rcpp::NumericVector values,
                            const double widthbefore,
                            const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  sma_next(values.begin(), times.begin(), &n, res.begin(),
           const_cast<double*>(&widthbefore), const_cast<double*>(&widthafter));
  return res;
}

//' @rdname SMAnext
// [[Rcpp::export]]
Rcpp::NumericVector SMAlast(Rcpp::DatetimeVector times,
                            Rcpp::NumericVector values,
                            const double widthbefore,
                            const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  sma_last(values.begin(), times.begin(), &n, res.begin(), 
           const_cast<double*>(&widthbefore), const_cast<double*>(&widthafter));
  return res;
}

//' @rdname SMAnext
// [[Rcpp::export]]
Rcpp::NumericVector SMAlinear(Rcpp::DatetimeVector times,
                              Rcpp::NumericVector values,
                              const double widthbefore,
                              const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  sma_linear(values.begin(), times.begin(), &n, res.begin(), 
           const_cast<double*>(&widthbefore), const_cast<double*>(&widthafter));
  return res;
}


src/rollingWrapper.cpp
#include <Rcpp.h>

extern "C" {
#include "rolling.h"
}

//' The UTS library by Andreas Eckner provides algorithms for unevenly
//' spaced time-series data.  This package brings a few of them to R.
//' The functions describe here offer various rolling operators.
//' @title Rolling operations functions for irregularly spaced time series
//' @param times A Datetime vector
//' @param values A numeric vector
//' @param widthbefore A double with the preceding observation width
//' @param widthafter A double with the subsequent observation width
//' @param moment A double with the requested moment.
//' @return A numeric vector with the corresponding result.
//' @author Dirk Eddelbuettel for the package, Andreas Eckner for the
//' underlying code.
// [[Rcpp::export]]
Rcpp::NumericVector rollingCentralMoment(Rcpp::DatetimeVector times,
                                         Rcpp::NumericVector values,
                                         const double widthbefore,
                                         const double widthafter,
                                         const double moment) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_central_moment(values.begin(), times.begin(), &n, res.begin(),
                         const_cast<double*>(&widthbefore),
                         const_cast<double*>(&widthafter),
                         const_cast<double*>(&moment));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingMax(Rcpp::DatetimeVector times,
                               Rcpp::NumericVector values,
                               const double widthbefore,
                               const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_max(values.begin(), times.begin(), &n, res.begin(),
              const_cast<double*>(&widthbefore),
              const_cast<double*>(&widthafter));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingMean(Rcpp::DatetimeVector times,
                                Rcpp::NumericVector values,
                                const double widthbefore,
                                const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_mean(values.begin(), times.begin(), &n, res.begin(),
               const_cast<double*>(&widthbefore),
               const_cast<double*>(&widthafter));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingMedian(Rcpp::DatetimeVector times,
                                  Rcpp::NumericVector values,
                                  const double widthbefore,
                                  const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_median(values.begin(), times.begin(), &n, res.begin(),
                 const_cast<double*>(&widthbefore),
                 const_cast<double*>(&widthafter));
  return res;
}
        
//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingMin(Rcpp::DatetimeVector times,
                               Rcpp::NumericVector values,
                               const double widthbefore,
                               const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_min(values.begin(), times.begin(), &n, res.begin(),
              const_cast<double*>(&widthbefore),
              const_cast<double*>(&widthafter));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingNobs(Rcpp::DatetimeVector times,
                                Rcpp::NumericVector values,
                                const double widthbefore,
                                const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_num_obs(values.begin(), times.begin(), &n, res.begin(),
                  const_cast<double*>(&widthbefore),
                  const_cast<double*>(&widthafter));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingProduct(Rcpp::DatetimeVector times,
                                   Rcpp::NumericVector values,
                                   const double widthbefore,
                                   const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_product(values.begin(), times.begin(), &n, res.begin(),
                  const_cast<double*>(&widthbefore),
                  const_cast<double*>(&widthafter));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingSD(Rcpp::DatetimeVector times,
                              Rcpp::NumericVector values,
                              const double widthbefore,
                              const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_sd(values.begin(), times.begin(), &n, res.begin(),
             const_cast<double*>(&widthbefore),
             const_cast<double*>(&widthafter));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingSum(Rcpp::DatetimeVector times,
                               Rcpp::NumericVector values,
                               const double widthbefore,
                               const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_sum(values.begin(), times.begin(), &n, res.begin(),
              const_cast<double*>(&widthbefore),
              const_cast<double*>(&widthafter));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingSumStable(Rcpp::DatetimeVector times,
                                     Rcpp::NumericVector values,
                                     const double widthbefore,
                                     const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_sum_stable(values.begin(), times.begin(), &n, res.begin(),
                     const_cast<double*>(&widthbefore),
                     const_cast<double*>(&widthafter));
  return res;
}

//' @rdname rollingCentralMoment
// [[Rcpp::export]]
Rcpp::NumericVector rollingVar(Rcpp::DatetimeVector times,
                               Rcpp::NumericVector values,
                               const double widthbefore,
                               const double widthafter) {
  if (times.size() != values.size()) Rcpp::stop("Matching vectors needed.");
  int n = times.size();
  Rcpp::NumericVector res(n);
  rolling_var(values.begin(), times.begin(), &n, res.begin(),
              const_cast<double*>(&widthbefore),
              const_cast<double*>(&widthafter));
  return res;
}


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' The UTS library by Andreas Eckner provides algorithms for unevenly
#' spaced time-series data.  This package brings a few of them to R.
#' The functions describe here offer exponentially-decaying weighted moving
#' average, or EMA, for short. Three variants are provides considering
#' the last or next observation relative to time \sQuote{t}, as well as
#' linear interpolation between them.
#' @title EMA functions for unevenly spaced time series
#' @param times A Datetime vector
#' @param values A numeric vector
#' @param tau A double with the decay factor
#' @return A numeric vector with EMA-weighted values.
#' package at the given position is available.
#' @author Dirk Eddelbuettel for the package, Andreas Eckner for the
#' underlying code.
#' @examples
#' if (requireNamespace("xts", quietly=TRUE)) {
#'     suppressMessages(library(xts))
#'     times <- ISOdatetime(2010, 1, 2, 8, 30, 0) + c(0, 1.0, 1.2, 2.3, 2.9, 5.0)
#'     values <- seq(0, 10, by=2)
#'     plot(xts(values, order.by=times), type="b",
#'          main="Series and last/next/linear EMAs",
#'          major.ticks="auto", grid.ticks.on="auto")
#'     lines(xts(EMAlast(times,values, 1),  values, order.by=times),
#'           type="b", col="lightblue")
#'     lines(xts(EMAnext(times,values, 1),  values, order.by=times),
#'           type="b", col="darkblue")
#'     lines(xts(EMAlinear(times,values, 1),  values, order.by=times),
#'           type="b", col="mediumblue")
#'     addLegend("topleft", legend.names=c("series", "EMAlast", "EMAnext", "EMAlinear"),
#'               lty=rep(1,4), lwd=rep(1,4),
#'               col=c("black", "lightblue", "darkblue", "mediumblue"))
#' }
EMAnext <- function(times, values, tau) {
    .Call(`_RcppUTS_EMAnext`, times, values, tau)
}

#' @rdname EMAnext
EMAlast <- function(times, values, tau) {
    .Call(`_RcppUTS_EMAlast`, times, values, tau)
}

#' @rdname EMAnext
EMAlinear <- function(times, values, tau) {
    .Call(`_RcppUTS_EMAlinear`, times, values, tau)
}

#' The UTS library by Andreas Eckner provides algorithms for unevenly
#' spaced time-series data.  This package brings a few of them to R.
#' The functions describe here offer various rolling operators.
#' @title Rolling operations functions for irregularly spaced time series
#' @param times A Datetime vector
#' @param values A numeric vector
#' @param widthbefore A double with the preceding observation width
#' @param widthafter A double with the subsequent observation width
#' @param moment A double with the requested moment.
#' @return A numeric vector with the corresponding result.
#' @author Dirk Eddelbuettel for the package, Andreas Eckner for the
#' underlying code.
rollingCentralMoment <- function(times, values, widthbefore, widthafter, moment) {
    .Call(`_RcppUTS_rollingCentralMoment`, times, values, widthbefore, widthafter, moment)
}

#' @rdname rollingCentralMoment
rollingMax <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingMax`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingMean <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingMean`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingMedian <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingMedian`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingMin <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingMin`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingNobs <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingNobs`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingProduct <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingProduct`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingSD <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingSD`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingSum <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingSum`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingSumStable <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingSumStable`, times, values, widthbefore, widthafter)
}

#' @rdname rollingCentralMoment
rollingVar <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_rollingVar`, times, values, widthbefore, widthafter)
}

#' The UTS library by Andreas Eckner provides algorithms for unevenly
#' spaced time-series data.  This package brings a few of them to R.
#' The functions describe here offer simple moving
#' average, or SMA, for short. Three variants are provides considering
#' the last or next observation relative to time \sQuote{t}, as well as
#' linear interpolation between them.
#' @title SMA functions for unevenly spaced time series
#' @param times A Datetime vector
#' @param values A numeric vector
#' @param widthbefore A double with the preceding observation width
#' @param widthafter gvA double with the subsequent observation width
#' @return A numeric vector with SMA-weighted values.
#' package at the given position is available.
#' @author Dirk Eddelbuettel for the package, Andreas Eckner for the
#' underlying code.
#' @examples
#' if (requireNamespace("xts", quietly=TRUE)) {
#'     suppressMessages(library(xts))
#'     times <- ISOdatetime(2018, 6, 7, 8, 30, 0) + c(0, 1.0, 1.2, 2.3, 2.9, 5.0)
#'     values <- seq(0, 10, by=2)
#'     plot(xts(values, order.by=times), type="b",
#'          main="Series and last/next/linear SMAs",
#'          major.ticks="auto", grid.ticks.on="auto")
#'     lines(xts(SMAlast(times,values, 2.5, 1), values, order.by=times), 
#'          type="b", col="lightblue")
#'     lines(xts(SMAnext(times,values, 2.5, 1),  values, order.by=times),
#'          type="b", col="darkblue")
#'     lines(xts(SMAlinear(times,values, 2.5, 1),  values, order.by=times),
#'          type="b", col="mediumblue")
#'     addLegend("topleft", legend.names=c("series", "SMAlast", "SMAnext", "SMAlinear"),
#'               lty=rep(1,4), lwd=rep(1,4),
#'               col=c("black", "lightblue", "darkblue", "mediumblue"))
#' }
SMAnext <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_SMAnext`, times, values, widthbefore, widthafter)
}

#' @rdname SMAnext
SMAlast <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_SMAlast`, times, values, widthbefore, widthafter)
}

#' @rdname SMAnext
SMAlinear <- function(times, values, widthbefore, widthafter) {
    .Call(`_RcppUTS_SMAlinear`, times, values, widthbefore, widthafter)
}

#' The UTS library by Andreas Eckner provides algorithms for unevenly
#' spaced time-series data.  This package brings a few of them to R.
#' This function shows the original example.
#' @title Irregularly spaced time series example
#' @return Nothing
#' @author Dirk Eddelbuettel for the package, Andreas Eckner for the
#' underlying code.
#' @examples
#' utsExample()
utsExample <- function() {
    invisible(.Call(`_RcppUTS_utsExample`))
}

